bundles/org.openhab.binding.senseenergy/src/main/java/org/openhab/binding/senseenergy/internal/handler/SenseEnergyBridgeHandler.java

```diff
@@ -0,0 +1,213 @@
+
+    private @Nullable SenseEnergyDiscoveryService discoveryService;
+
+    private SenseEnergyApi api;
+    private SenseEnergyBridgeConfiguration config;
+
+    protected @Nullable ScheduledFuture<?> heartbeatJob;
+
+    private Set<Long> monitorIDs = Collections.emptySet();
+
+    public SenseEnergyBridgeHandler(final Bridge thing, HttpClient httpClient) {
+        super(thing);
+        api = new SenseEnergyApi(httpClient);
+
+        config = getConfigAs(SenseEnergyBridgeConfiguration.class);
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Set.of(SenseEnergyDiscoveryService.class);
+    }
+
+    public SenseEnergyApi getApi() {
+        return api;
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        scheduler.execute(this::goOnline);
+    }
+
+    public void goOnline() {
+        try {
+            this.monitorIDs = api.initialize(config.email, config.password);
+        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+            handleApiException(e);
+            return;
+        }
+
+        refreshMonitors();
+
+        updateStatus(ThingStatus.ONLINE);
+
+        this.heartbeatJob = scheduler.scheduleWithFixedDelay(this::heartbeat, 0, HEARTBEAT_MINUTES, TimeUnit.MINUTES);
+    }
+
+    private void heartbeat() {
+        ThingStatus thingStatus = getThing().getStatus();
+
+        if (thingStatus == ThingStatus.OFFLINE
+                && getThing().getStatusInfo().getStatusDetail() == ThingStatusDetail.COMMUNICATION_ERROR) {
+            goOnline(); // only attempt to goOnline if not a configuration error
+            return;
+        }
+
+        if (thingStatus != ThingStatus.ONLINE) {
+            return;
+        }
+
+        // token is verified on each api call, called here in case no API calls are made in the alloted period
+        try {
+            getApi().verifyToken();
+        } catch (InterruptedException | TimeoutException | ExecutionException | SenseEnergyApiException e) {
+            handleApiException(e);
+        }
+
+        // call heartbeat for each thing to check health
+        getThing().getThings().stream() //
+                .map(t -> t.getHandler()) //
+                .filter(h -> h instanceof SenseEnergyMonitorHandler) //
+                .map(h -> (SenseEnergyMonitorHandler) h) //
+                .forEach(h -> h.heartbeat());
+    }
+
+    public void handleApiException(Exception e) {
+        ThingStatusDetail statusDetail = ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR;
+
+        if (e instanceof SenseEnergyApiException apiException) {
+            statusDetail = apiException.isConfigurationIssue() ? ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR
+                    : ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR;
+        } else {
+            logger.warn("Unhandled Exception", e);
```

Comment:
the stacktrace can be logged to debug. The thing state is already set, so no need to have it at warn level.
```suggestion
            logger.debug("Unhandled Exception", e);
```